tools/github_update.py

```diff
@@ -96,118 +104,235 @@ def _get_repo_path(self):
+        if state == PENDING_STATE:
+            start_time_file = open(start_time_path, 'w')
+            start_time_file.write('{}\n'.format(str(time.time())))
+            start_time_file.flush()
+            start_time_file.close()
+        elif os.path.isfile(start_time_path):
+            try:
+                start_time = float(open(start_time_path).read().strip())
+                message = '[{}] {}'.format(self._pretty_duration(time.time() - start_time), message)
+                os.remove(start_time_path)
+            except:
+                message = '[?] {}'.format(message)
+
+        request = self._build_post_update_request(context_label, state, message, details_url)
+        response = self._send_request(request)
+        if not self._check_success(response, request, 'status update'):
             return
-        logger.info('Updated GitHub PR with status: {}'.format(request['path']))
+
+
+def _get_details_link_url():
+    '''returns the url to be included as the details link in the status'''
+    details_url = os.environ.get('GITHUB_COMMIT_STATUS_URL', '') # custom URL via env
+    if not details_url:
+        details_url = os.environ.get('BUILD_URL', '') # provided by jenkins
+        if details_url:
+            details_url += 'console'
+    return details_url
+
+
+def _get_auth_token():
+    github_token = os.environ.get('GITHUB_TOKEN_REPO_STATUS', '')
+    if not github_token:
+        github_token = os.environ.get('GITHUB_TOKEN', '')
+    if not github_token:
+        raise Exception(
+            'Failed to determine auth token to use with GitHub. ' +
+            'Provide either GITHUB_TOKEN or GITHUB_TOKEN_REPO_STATUS in env.')
+    return github_token
+
+
+def _should_access_github():
+    if not 'WORKSPACE' in os.environ:
+        # not running in CI. skip actually sending anything to GitHub
+        return False
+    if os.environ.get('GITHUB_DISABLE', ''):
+        # environment has notifications disabled. skip actually sending anything to GitHub
+        return False
+    if not (os.environ.get('GITHUB_COMMIT_STATUS_URL') or os.environ.get('BUILD_URL')):
+        # CI job was not triggered by a PR
+        return False
+    return True
 
 
 def print_help(argv):
-    logger.info('Syntax: {} <state: pending|success|error|failure> <context_label> <status message>'.format(argv[0]))
+    logger.info('Syntax:')
+    logger.info('- Update: {} <state: {}> <context_label> [a status message ...]'.format(
+        argv[0], '|'.join(VALID_STATES)))
+    logger.info('- Reset pending: {} reset [a replacement message ...]'.format(argv[0]))
 
 
 def main(argv):
-    if len(argv) < 4:
+    if len(argv) < 2:
         print_help(argv)
         return 1
+
+    updater = GithubStatusUpdater()
     state = argv[1]
-    if state != 'pending' \
-            and state != 'success' \
-            and state != 'error' \
-            and state != 'failure':
+    if state == 'reset':
```

Comment:
For readability, can you factor each of these two stanzas into its own function?